fix(evm): revert precompile

MaxMustermann2 opened the pull request:
needs dependencies to be merged first

Step 1 https://github.com/imua-xyz/cosmos-sdk/pull/1

Step 2 https://github.com/imua-xyz/evmos/pull/7

Step 3 This PR

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Chores**
  - Updated dependency sources to enhance functionality and support for new features.
  - Adjusted configuration values, including gateway addresses and token thresholds, to ensure smoother local operations.

- **Refactor**
  - Streamlined backend processing for transaction execution and state management, enhancing reliability and performance.
  - Introduced dedicated routines for clearing outdated state data, supporting more robust account operations.
  - Enhanced error handling and input validation in various precompile methods to improve robustness.
  - Added new methods for managing state and code deletions, improving code modularity.

- **New Features**
  - Introduced new smart contracts (`Gateway`, `GatewayCaller`, `GatewayCallee`, `UnknownMethodCaller`, `PrecompileCallerThatReverts`, `ThirdPartyCallee`, and `TryCatchCaller`) with functionalities for asset management, revert handling, and error capturing.
  - Added JSON files for ABI and binary code of new contracts, facilitating interaction with the blockchain.
  - Implemented a new function for validating transaction methods, enhancing error handling during execution.

- **Tests**
  - Added new test suite for validating wrapped revert scenarios in smart contracts, ensuring robust error handling during asset operations.
  - Added a new test suite for testing unknown method calls on precompiles, ensuring graceful handling of such scenarios.
  - Improved error handling and state management in existing test cases for better reliability.
  - Enhanced log checking with expected data validation for emitted events.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

MaxMustermann2 committed:
fix(evm): revert precompile

needs dependencies to be merged first

TimmyExogenous reviewed (approved):
LGTM

MaxMustermann2 committed:
fix(go-mod): update remote commit dep

MaxMustermann2 committed:
fix(evm): check input len, optional journal entry

MaxMustermann2 committed:
fix(go-mod): point to new commit

MaxMustermann2 committed:
fix(precompiles): use `Addr` not `SetAddress`

MaxMustermann2 committed:
fix(tests): pass precompile tests with Commit()

MaxMustermann2 committed:
tests(precompile): add wrapped revert test

MaxMustermann2 commented on precompiles/assets/assets.go:
```diff
@@ -183,6 +177,10 @@ func (p Precompile) Run(evm *vm.EVM, contract *vm.Contract, readOnly bool) (bz [
+	if err := p.AddJournalEntries(stateDB, snapshot); err != nil {
+		return nil, err
+	}
+
```
We can examine two scenarios in which this limit may be triggered by a caller.
1. Wrapped in a try catch: added unit test for such a situation
2. Not wrapped in a try catch: the overall tx fails and hence it is never committed to state

MaxMustermann2 committed:
test(precompile): explain why it works

- When adding a precompile journal entry, the latest entry is first
  added and then the limit is checked.
- If the limit is breached, the latest entry is then reverted. Hence,
  the impact of the limit overage is reverted.
- Even if this reversion is wrapped in a try/catch, it is thus not
  written to state

MaxMustermann2 committed:
doc(tests): update documentation to explain test

adu-web3 commented on precompiles/assets/assets.go:
```diff
@@ -86,9 +85,9 @@ func (p Precompile) Run(evm *vm.EVM, contract *vm.Contract, readOnly bool) (bz [
-	}
+	// we cache the context because we consume the failure, if any, in this function.
+	// cascading it to statedb is possible but we will lose foundry compatibility
+	// see https://github.com/imua-xyz/imuachain/issues/70
```
#70 is not about foundry compatibility, so we can remove this comment

adu-web3 commented on precompiles/assets/assets.go:
```diff
@@ -86,9 +85,9 @@ func (p Precompile) Run(evm *vm.EVM, contract *vm.Contract, readOnly bool) (bz [
+	// we cache the context because we consume the failure, if any, in this function.
+	// cascading it to statedb is possible but we will lose foundry compatibility
+	// see https://github.com/imua-xyz/imuachain/issues/70
 	cc, writeFunc := ctx.CacheContext()
```
both evmos's patch and cronos's implementation seems not having this, does it mean it is safe to remove this?

MaxMustermann2 replied:
Added comment to indicate that it will be done in the next PR.

adu-web3 commented on precompiles/assets/assets.go:
```diff
@@ -183,6 +178,16 @@ func (p Precompile) Run(evm *vm.EVM, contract *vm.Contract, readOnly bool) (bz [
 		return nil, vm.ErrOutOfGas
 	}
 
+	if p.IsTransaction(method.Name) {
```
this `isTransaction` is not safe: if some newly added method would write the store but is forgotten to be declared in `p.IsTransaction`, this function would return false. Given `isTransaction` is enforced by precompile interface, could we add a new function like `safeIsTransaction` that would return an error if it is not declared?

MaxMustermann2 replied:
Changed `IsTransaction` to `panic` in the `default` situation and captured this `panic` in `init` to force developers to remember this.

adu-web3 commented on precompiles/assets/testdata/Gateway.sol:
```diff
@@ -0,0 +1,65 @@
+        uint256 opAmount
+    ) public returns (bool success, uint256 latestAssetState) {
+        (success, latestAssetState) = withdrawLST(clientChainID, assetsAddress, withdrawAddress, opAmount);
+        revert("Intentional revert after withdraw");
```
The scenario that reflects the problem this time is not a direct revert following the precompile call, but calling another contract after calling the precompile and revert in this outcome call. So instead of directly reverting here, we should call another contract here and make that call revert

MaxMustermann2 replied:
Gateway is not called in the unit test. Instead, GatewayCaller calls gateway.

adu-web3 replied:
Yeah, I mean the old node would not trigger the error if we directly revert after calling the precompile in the same function, unless we call another contract and revert there, according to my test

MaxMustermann2 replied:
Updated so that a callee contract is responsible for the revert instead.

adu-web3 commented:
btw, would we also patch the balance discrepancy issue in this PR, or in `statedb` of our evmos fork? And it may not be related to this, but Cronos's implementation inspires us that we could convert native event logs emitted in modules as `ethlogs` to be emitted in evm, which should help improve contract development, I could open another PR to see if this works

MaxMustermann2 committed:
fix(precompile): add IsTransaction validate

MaxMustermann2 committed:
tests(precompile): fix panic case

MaxMustermann2 committed:
fix(precompile): check panic for unknown method

MaxMustermann2 committed:
fix(test): revert in different contract

MaxMustermann2 committed:
fix(precompile): sort methods by string

MaxMustermann2 commented on precompiles/common/is_tx.go:
```diff
@@ -0,0 +1,67 @@
+	// although this method is only used during `init()` currently,
+	// this is a form of future-proofing.
+	methods := make([]string, 0, len(abi.Methods))
+	for _, method := range abi.Methods {
+		methods = append(methods, method.Name)
+	}
```
This iteration is performed to sort the map's keys. The sorted keys are then iterated over.

MaxMustermann2 committed:
test(precompiles): add `IsTransaction` panic test

- `p.IsTransaction` has been designed to `panic` whenever an unknown
  method name is provided.
- If the gateway (or anyone) calls the precompile with an unknown
  method, the node does not panic.
- This is because `IsTransaction` is only called after validating the
  method name.
- The added unit test validates this for the assets precompile.

MaxMustermann2 committed:
doc(app): add TODO for unit test

MaxMustermann2 committed:
fix(precompiles): add unknown method test

MaxMustermann2 committed:
Merge upstream/develop

MaxMustermann2 committed:
Merge upstream/develop

MaxMustermann2 committed:
go-dep: use transient storage dep

MaxMustermann2 committed:
fix(evm): change dep

MaxMustermann2 committed:
test(evm): demonstrate adu's precompile test

MaxMustermann2 committed:
fix(test): record failure count too

MaxMustermann2 committed:
fix(evm): update test

MaxMustermann2 committed:
fix(evm): update dependency and test

MaxMustermann2 committed:
test(evm): update values

MaxMustermann2 commented on precompiles/assets/testdata/PrecompileCallerThatReverts.sol:
```diff
@@ -0,0 +1,214 @@
+    }
+
+    // Helper function to register a token
+    function _registerOrUpdateToken(
+        uint8 clientChainId,
+        bytes memory token,
+        uint8 decimals,
+        string memory name,
+        string memory metadata,
+        string memory oracleInfo
+    ) internal {
+        uint32 clientChainIdUint32 = uint32(clientChainId);
+        bool registered = ASSETS_CONTRACT.registerToken(
+            clientChainIdUint32, 
+            token, 
+            decimals, 
+            name, 
+            metadata, 
+            oracleInfo
+        );
+        
+        if (!registered) {
+            bool updated = ASSETS_CONTRACT.updateToken(clientChainIdUint32, token, metadata);
+            require(updated, "Failed to update token");
+        }
+    }
```
Test; not relevant.

MaxMustermann2 commented on precompiles/assets/testdata/PrecompileCallerThatReverts.sol:
```diff
@@ -0,0 +1,214 @@
+    }
+
+    // Helper function to register a client chain
+    function _registerOrUpdateClientChain(
+        uint8 clientChainId,
+        uint8 stakerAccountLength,
+        string memory name,
+        string memory metadata,
+        string memory signatureScheme
+    ) internal {
+        (bool success,) = ASSETS_CONTRACT.registerOrUpdateClientChain(
+            uint32(clientChainId), 
+            stakerAccountLength, 
+            name, 
+            metadata, 
+            signatureScheme
+        );
+        
+        require(success, "Failed to register client chain");
+    }
```
Test; not relevant.

MaxMustermann2 commented on precompiles/assets/testdata/PrecompileCallerThatReverts.sol:
```diff
@@ -0,0 +1,214 @@
+        bytes calldata staker,
+        uint256 amount,
+        uint256 gasLimit
+    ) public returns (bool success) {
+        nonce += 1;
+        bytes memory data = abi.encodeWithSelector(
+            ASSETS_CONTRACT.depositLST.selector,
+            clientChainID,
+            token,
+            staker,
+            amount
+        );
+        address target = address(ASSETS_CONTRACT);
+        assembly {
+            let ptr := add(data, 0x20)
+            success := call(
+                gasLimit, target, 0, add(data, 0x20), mload(data), ptr, 0x20
+            )
+            success := and(success, mload(ptr))
+        }
```
Removed function `callPrecompileAndRevertGasStarved`

MaxMustermann2 committed:
fix(evm): respond to AI comments

MaxMustermann2 committed:
refactor(precompiles): remove repetitive code

- No need to check len(Input) < 4 since MethodByID does the same check
- Logging statement can be made common across the board
- Ensure atomicity within cache context as far as possible

adu-web3 commented:
all e2e test cases have passed, excellent work!

MaxMustermann2 committed:
fix(precompiles): respond to AI comments

MaxMustermann2 commented on precompiles/slash/slash.go:
```diff
@@ -112,16 +101,21 @@ func (p Precompile) Run(evm *vm.EVM, contract *vm.Contract, readOnly bool) (bz [
-		return false
+		panic(fmt.Sprintf("unknown method: %s", methodName))
+	}
+}
```
The `init` takes care of this. However, I have added a comment as you advised referring to the upstream guarantees.

MaxMustermann2 commented on precompiles/bls/bls.go:
```diff
@@ -45,21 +44,19 @@ func NewPrecompile() (*Precompile, error) {
 	if err != nil {
 		return 0
 	}
+	// TODO: @devin: add gas calculation depending on the method
```
Removed TODO since it was already implemented.

MaxMustermann2 commented on precompiles/bech32/bech32.go:
```diff
@@ -50,53 +51,64 @@ func NewPrecompile(authzKeeper authzkeeper.Keeper) (*Precompile, error) {
-	// bug fix to commit dirty objects
-	if err := stateDB.Commit(); err != nil {
+
+	args, err := method.Inputs.Unpack(contract.Input[4:])
+	if err != nil {
 		return nil, err
 	}
 
 	switch method.Name {
 	case MethodHexToBech32:
-		return p.HexToBech32(method, args)
+		bz, err = p.HexToBech32(method, args)
 	case MethodBech32ToHex:
-		return p.Bech32ToHex(method, args)
+		bz, err = p.Bech32ToHex(method, args)
+	default:
+		// should never happen
+		bz, err = nil, fmt.Errorf(cmn.ErrUnknownMethod, method.Name)
 	}
 
-	cost := ctx.GasMeter().GasConsumed() - initialGas
-	if !contract.UseGas(cost) {
-		return nil, vm.ErrOutOfGas
+	if err != nil {
+		// this will (might?) cause the entire tx to fail.
+		// it is acceptable because it would represent an error
+		// in the precompile caller.
+		return nil, err
 	}
-	return nil, nil
+
+	return bz, nil
 }
```
Pointless suggestion given the interface guard `var _ vm.PrecompiledContract = &Precompile{}` on top of the file.

MaxMustermann2 committed:
fix(ci): allow goreleaser to work in merge groups
